app/helpers/automated_tests_helper.rb

```diff
@@ -380,6 +474,435 @@ def parse_test_output(repo_assignment_dir, assignment, logfile, data)
+
+      # TODO: handle this error better
+      raise "error"
+    else
+      process_result(result, grouping_id, @assignment.id)
+    end
+
+  end
+
+  # From a list of test servers, choose the next available server
+  # using round-robin. Return the id of the server, and return -1
+  # if no server is available.
+  # TODO: keep track of the max num of tests running on a server
+  def self.choose_test_server()
+
+    if (defined? @last_server) && MarkusConfigurator.automated_testing_engine_on?
+      # find the index of the last server, and return the next index
+      @last_server = (@last_server + 1) % MarkusConfigurator.markus_ate_num_test_servers
+    else
+      @last_server = 0
+    end
+
+    return @last_server
+  end
+
+
+  # Launch the test on the test server by scp files to the server
+  # and run the script.
+  # This function returns two values: first one is the output from
+  # stdout or stderr, depending on whether the execution passed or
+  # had error; the second one is a boolean variable, true => execution
+  # passeed, false => error occurred.
+  def self.launch_test(server_id, assignment, repo_dir, call_on)
+    # Get src_dir
+    src_dir = File.join(repo_dir, assignment.repository_folder)
+
+    # Get test_dir
+    test_dir = File.join(MarkusConfigurator.markus_config_automated_tests_repository, assignment.repository_folder)
+
+    # Get the name of the test server
+    server = @list_of_servers[server_id]
+
+    # Get the directory and name of the test runner script
+    test_runner = MarkusConfigurator.markus_ate_test_runner_script_name
+
+    # Get the test run directory of the files
+    run_dir = MarkusConfigurator.markus_ate_test_run_directory
+
+    # Delete the test run directory to remove files from previous test
+    stdout, stderr, status = Open3.capture3("ssh #{server} rm -rf #{run_dir}")
+    if !(status.success?)
+      return [stderr, false]
+    end
+
+    # Recreate the test run directory
+    stdout, stderr, status = Open3.capture3("ssh #{server} mkdir #{run_dir}")
+    if !(status.success?)
+      return [stderr, false]
+    end
+
+    # Securely copy source files, test files and test runner script to run_dir
+    stdout, stderr, status = Open3.capture3("scp -p -r '#{src_dir}'/* #{server}:#{run_dir}")
+    if !(status.success?)
+      return [stderr, false]
+    end
+
+    stdout, stderr, status = Open3.capture3("scp -p -r '#{test_dir}' #{server}:#{run_dir}")
+    if !(status.success?)
+      return [stderr, false]
+    end
+
+    stdout, stderr, status = Open3.capture3("ssh #{server} cp #{test_runner} #{run_dir}")
+    if !(status.success?)
+      return [stderr, false]
+    end
+
+    # Find the test scripts for this test run, and parse the argument list
+    list_run_scripts = scripts_to_run(assignment, call_on)
+    arg_list = ""
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.